sst/SimengCoreWrapper.cc

```diff
@@ -0,0 +1,218 @@
+#include <sst/core/sst_config.h>
+#include "SimengCoreWrapper.hh"
+
+#include <iostream>
+
+// #include "simeng/RegisterValue.hh"
+
+using namespace SST::SSTSimeng;
+using namespace SST::Interfaces;
+// using namespace simeng;
+
+enum class SimulationMode { Emulation, InOrderPipelined, OutOfOrder };
+
+SimengCoreWrapper::SimengCoreWrapper(SST::ComponentId_t id, SST::Params& params): SST::Component(id) {
+    output.init("SimengCoreWrapper[" + getName() + ":@p:@t]:", 999, 0, SST::Output::STDOUT);
+    clock = registerClock(params.find<std::string>("clock", "1GHz"), new SST::Clock::Handler<SimengCoreWrapper>(this, &SimengCoreWrapper::clockTick));
+
+    executable_path = params.find<std::string>("executable_path", "");
+    executable_args = params.find<std::string>("executable_args", "");
+    config_path = params.find<std::string>("config_path", "");
+    cache_line_width  = params.find<uint64_t>("cache_line_width", "64");
```

Comment:
AFAIK you can't have multiple configs in SST. To run a simulation the command `sst config.py` is used. So i'm sure how we can pass multiple configs. I don't think SST can exeucte python logic so we can programtically import it inside `config.py`. However we can write a python script which generates a config.py per specification.

Reply:
I didn't mean to suggest we have multiple configs. The additional config file was in addition to the SimEng config file. So we'd have a config file for the core and a config file for the memory subsystem. At the moment, you have SST config params as CLI args, but, if there are many possible params we'd like to target, then the amount of CLI args may become cumbersome. It may be better to pass a config file and extract SST args from that (similar to how we do for SimEng).

Reply:
I'm not sure if I understand you correctly. These config params aren't passed as CLI args, but rather these are passed inside the `config.py` file. Since `SimengCoreWrapper` is an SST component, we identify it in `config.py` by the tag `sstsimeng.simengcore` and then pass parameters to it. Those params are then exposed to us via the params object provided by SST. The only command I run is `sst config.py`. 

```
// sst/config/config.py
cpu = sst.Component("core", "sstsimeng.simengcore")
cpu.addParams({
    "config_path": "/home/rahat/asimov/SimEng/configs/a64fx.yaml",
    "executable_path": "/home/rahat/asimov/simeng-progs/stream",
    "executable_args": "",
    "clock" : "1GHz",
    "max_addr_memory": 1*1024*1024*1024-1,
    "cache_line_width": "64",
})

```
By passing params like this we can also document then via ELI Macros provided by SST:
```
// sst/include/SimengCoreWrapper.hh

class SimengCoreWrapper : public SST::Component {
 ......
  SST_ELI_DOCUMENT_PARAMS(
        { "config_path", "Path to Simeng YAML config file (string)", "" },
        { "executable_path", "Path to executable binary to be run by SimEng (string)", "" },
        { "executable_args", "argument to be passed to the executable binary (string)", "" },
        { "clock", "Clock rate of the SST clock (string)", "" },
        { "max_addr_memory", "Maximum address that memory can access (int)"},
 )
 ........
}
```
Then when we get all information about our custom SST::Component by using the command `sst-info --libs=sstsimeng`:
```
Filtering output on Element = "sstsimeng"
================================================================================
ELEMENT 0 = sstsimeng ()
Num Components = 1
      Component 0: simengcore
      CATEGORY: PROCESSOR COMPONENT
         NUM STATISTICS = 0
         NUM PORTS = 0
         NUM SUBCOMPONENT SLOTS = 0
         NUM PARAMETERS = 5
            PARAMETER 0 = config_path (Path to Simeng YAML config file (string)) []
            PARAMETER 1 = executable_path (Path to executable binary to be run by SimEng (string)) []
            PARAMETER 2 = executable_args (argument to be passed to the executable binary (string)) []
            PARAMETER 3 = clock (Clock rate of the SST clock (string)) []
            PARAMETER 4 = max_addr_memory (Maximum address that memory can access (int)) [<required>]
    simengcore: Simeng core wrapper for SST
    Using ELI version 0.9.0
    Compiled on: Aug 10 2022 10:36:49, using file: /home/rahat/asimov/SimEng/sst/include/SimengCoreWrapper.hh
Num SubComponents = 0
Num Modules = 0
Num SSTPartitioners = 0

```
Is this not the right way to do it? 
Or worse have I completely misunderstood you again?

Reply:
Okay no, this is good, I understand now. So we'd have to alter this config.py file when we want to alter the memory hierarchy we are simulating within SST?

Reply:
Yes only this file.